lib/file/atomic.go

```diff
@@ -0,0 +1,51 @@
+package file
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/go-uuid"
+)
+
+// WriteAtomic writes the given contents to a temporary file in the same
+// directory, does an fsync and then renames the file to its real path
+func WriteAtomic(path string, contents []byte) error {
+	return WriteAtomicWithPerms(path, contents, 0700, 0600)
+}
+
+func WriteAtomicWithPerms(path string, contents []byte, dirPerms, filePerms os.FileMode) error {
+
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return err
+	}
+	tempPath := fmt.Sprintf("%s-%s.tmp", path, uuid)
+
+	if err := os.MkdirAll(filepath.Dir(path), dirPerms); err != nil {
+		return err
+	}
+	fh, err := os.OpenFile(tempPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, filePerms)
+	if err != nil {
+		return err
+	}
```

Comment:
The stdlib has methods for everything we're doing here that use platform-specific implementations under the hood. Let's use those so we know we'll work everywhere.

Reply:
Hey @tgross,
Thanks for all the great comments and support. 
I am not sure what I have missed here as the os package is part of the standard library 
https://pkg.go.dev/os@go1.19.1#pkg-index
It would be great to know where to start thanks

Reply:
Sorry, I should have called out specifically what I meant: [`os.CreateTemp`](https://pkg.go.dev/os#CreateTemp). And you can do a `defer os.RemoveAll` on the resulting file path to ensure it's always removed on any return, if it still exists.

Reply:
I have been looking at this and os.CreateTemp does not allow you to write the file with permissions already set. We need specific permissions on the key file as it is a sensitive file.
I have removed the duplicated os.Remove(tempPath)